sound/soc/sof/debug.c

```diff
@@ -17,6 +17,167 @@
 #include "sof-priv.h"
 #include "ops.h"
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_IPC_FLOOD_TEST)
+#define MAX_IPC_FLOOD_DURATION_MS 1000
+#define MAX_IPC_FLOOD_COUNT 10000
+
+static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
+				    bool flood_duration_test,
+				    unsigned long ipc_duration_ms,
+				    unsigned long ipc_count)
+{
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_reply reply;
+	s64 min_response_time = LONG_MAX;
+	ktime_t start, end, test_end;
+	s64 avg_response_time = 0;
+	s64 max_response_time = 0;
+	s64 ipc_response_time;
+	int i = 0;
+	int ret;
+
+	/* configure test IPC */
+	hdr.cmd = SOF_IPC_GLB_TEST | SOF_IPC_TEST_IPC_FLOOD;
+	hdr.size = sizeof(hdr);
+
+	/* set test end time for duration flood test */
+	if (flood_duration_test)
+		test_end = ktime_get_ns() + ipc_duration_ms * 1000000;
+
+	/* send test IPC's */
+	while (1) {
+		start = ktime_get();
+		ret = sof_ipc_tx_message(sdev->ipc, hdr.cmd, &hdr, hdr.size,
+					 &reply, sizeof(reply));
+		end = ktime_get();
+
+		if (ret < 0)
+			break;
+
+		/* compute min and max response times */
+		ipc_response_time = ktime_to_ns(ktime_sub(end, start));
+		min_response_time = min(min_response_time, ipc_response_time);
+		max_response_time = max(max_response_time, ipc_response_time);
+
+		/* calculate the moving average */
+		avg_response_time +=
+			(ipc_response_time - avg_response_time) / (i + 1);
```

Comment:
does this work in i386 mode? I thought we'd need to use a long division?

Reply:
also the precision decrease with the number of interactions, why not just do the raw sum and divide last to avoid introducing errors in the count. Implementing a mean with an IIR is asking for trouble.